README.md
# PANPRS

### Installation

For the sparse matrix implementation, please install the R package as follows:

```R
devtools::install_github("katherine-h-l/PANPRSnext@sparse", force = TRUE)
```

For the dense matrix implementation, please install the R package as follows:

```R
devtools::install_github("katherine-h-l/PANPRSnext@master", force = TRUE)
```

### Input for PANPRS incorporating multiple traits and functional annotations of SNPs.

summaryZ,
The Z statistics of p SNPs from q GWA studies. A matrix with dimension p x q for p SNPs and q traits. The first column corresponds to the primary trait and the rest columns correspond to the secondary traits.

Nvec,
A vector of length q for the sample sizes of q GWA studies.

plinkLD,
LD matrix information.

NumIter,
The number of maximum iterations for the estimation procedure.

funcIndex,
Inputs for the functional annotations of SNPs. A p x k matrix with (0,1) entry; p is the number of SNPs and
k is the number of functional annotations. For the element at i-th row, j-th column, the entry 0 means SNP i without j-th functional annotation; entry 1 means otherwise.

numfunc,
The number of functional annotations.

dfMax
The upper bound of the number of non-zero estimates of coefficients for the primary trait.

### Usage:

The current version only work on Unix, Linux and Mac System, R(>=3.4.3), R packages "gtools" and "permutations" and GCC(>=4.4.7) are required.

Modify the parameters in the gsfPEN.R, and execute it.

## Example

Please find it in the R package.


CHANGELOG.md
# Changes

-   Finished Armadillo implementation of gsPEN and gsfPEN

    -   New File Structure: PANPRS_gsfPEN.cpp PANPRS_gsfPEN.hpp PANPRS_gsfPEN.R, PANPRS_gsPEN.cpp PANPRS_gsPEN.hpp PANPRS_gsPEN.R
    -   Description: Fixed a bug in PANPRS_gsfPEN.cpp that caused matrix values to be updated incorrectly. Fixed the gsPEN implementation using Armadillo. Both the Armadillo gsPEN and gsfPEN implementations work as expected and give identical results to the original package now.

-   Merged rcpp branch into the master branch

    -   The package can now be installed directly from github using:

    ```r
    devtools::install_github("katherine-h-l/PANPRS-next", force = TRUE)
    ```

    -   Need to update gsPEN.R and gsPEN.cpp now as well.
        -   Renamed and created new files.
        -   PANPRS.R -> PANPRS_gsfPEN.R
        -   PANPRS.cpp -> PANPRS_gsfPEN.cpp
        -   PANPRS.hpp -> PANPRS_gsfPEN.hpp
        -   New: PANPRS_gsPEN.R, PANPRS_gsPEN.cpp, PANPRS_gsPEN.hpp

-   Increased peroformance greatly

    -   Modified files: PANPRS.cpp
    -   Description: Used branchless programming to replace many if statements allowing the compiler to load the instructions into the CPU pipeline more efficiently. Further improved performance by replacing many for loops with vectorized operations using armadillo functions.

-   Increased performance by elimating unessary overwrites to the beta matrix

    -   Modified files: PANPRS.cpp
    -   Description: Simply adding an if statement on line 283 to check if we are setting an element of a sparse matrix to 0 (which has no effect) increased performance by up to 30%.

-   Ran additional memory profiling on larger data sets with the new sparse matrix implementation.

    -   New directory: MemoryProfiling
    -   Description: Using profvis, we are able to profile the code showing when and where memory is being allocated / deallocated. Not only does the package use ~50% less memory while performing and executing the code, it also returns much smaller objects. This reduction in the size of the returned objects is once again due to the use of sparse matrices, and we are saving upwards of 93% of the memory that would have been used by the previous version of the package.
    -   TO DO: We can reduce the memory usage even further by reducing the number of temporary allocations. the temp_beta_matrix and joint_beta_matrix are both potential candidates for removal.

-   Implemented sparse matrix storage for return value beta_matrix.

    -   Modified files: PANPRS.cpp
    -   Description: Reduced the memory usage of the package drastically at the cost of some performance. The package now uses sparse matrices to store the beta matrix. This reduces the memory usage of the package by ~50% and allows the package to run on larger datasets.
    -   Caveat: The package is now ~2x slower than the previous version. This is due to the overhead of using a sparse matrix. Accessing elements of sparse matrices comes with a performance penalty. This is a tradeoff between memory usage and performance. We can explore further performance improvements to reclaim some of the lost performace.

-   Package compilation has been test successfully tested on Linux (Manjaro) and Windows 11.

    -   The package fails to compile on Mac OS using Apple M2 Arm chip due to failure to link to fortran compiler. I suspect this is an issue with the clang compiler. To be resolved still.

-   TO DO:

    -   Have yet to leverage sparse matrices. We already have decreased memory usage just from transitioning to Armadillo and cleaning up some memory allocations. Sparse matrices should be able to further reduce memory usage.
        -   What matrices should be treated as sparse?
    -   Further reduce the amount of temporary allocations. Try and find a way to elimate the temporary matrices entirely.
    -   Implement parallelization using OpenMP
        -   What parts of the code should be parallelized?
    -   Attempt to convert loops to a branchless style. This should reduce the overhead and increase performance of the for loops even further.

-   Ran some memory profiling on the code. The new version of the code uses approximately ~17% less memory already.

    -   Modified files: Test.Rmd
    -   Output files: Test.html new.html old.html
    -   Description: Files show when and how much memory is allocated in each part of the gsfpen function calls.

-   Fixed bug in the C++ code that was causing the results to be incorrect. Zmatrix was not being initialized correctly.

    -   Modified files: PANPRS.cpp
    -   How error was found: Took a subset of the initial summaryZ and funcIndex data. Reduced the number of tuning parameters as well so that the output would be printable / readable. Inserted various print statements through the loops in order to determine where values were different accross the new and old code.
    -   Description: Zmatrix needed to be converted from data frame to a matrix to be passed to C++. My previous implementation was incorrect which has now been fixed. Additionally, there was a bug in the R Internal functions that was causing the lambda vectors to be initialized incorrectly. This has also been fixed.
    -   Caveat: The code now matches the previous version of the code in terms of output. Now that all bugs have been fixed though, the new code is now slightly slower (by ~25%) than the previous version. However, we now have everything working with Armadillo and can move to optimization.

-   Fixed calculations in new version of the code. Issue was within one of the loops in the C++ code.

    -   Modified files: PANPRS.cpp
    -   How error was found: Took a subset of the initial summaryZ and funcIndex data. Reduced the number of tuning parameters as well so that the output would be printable / readable. Inserted various print statements through the loops in order to determine where values were different accross the new and old code.

R/PANPRS.R
#' Run the package on the provided data set
#' @export
test_pkg <- function(
    debug_output = FALSE) {
  data("summaryZ")
  data("Nvec")
  data("plinkLD")
  data("funcIndex")
  output <- gsfPEN_R(
    summary_z = summaryZ, # nolint: object_usage_linter.
    n_vec = Nvec, # nolint: object_usage_linter.
    plinkLD = plinkLD, # nolint: object_usage_linter.
    func_index = funcIndex, # nolint: object_usage_linter.
    debug_output = debug_output
  )

  return(output)
}

#' Run the package on the provided data set
#' DEBUG VERSION
#' @export
debug_test_pkg <- function(
    summaryZ,
    funcIndex,
    debug_output = FALSE,
    sub_tuning = 1,
    lambda_vec_limit_len = c(1.5, 1)) {
  data("summaryZ")
  data("Nvec")
  data("plinkLD")
  data("funcIndex")
  output <- gsfPEN_R(
    summary_z = summaryZ, # nolint: object_usage_linter.
    n_vec = Nvec, # nolint: object_usage_linter.
    plinkLD = plinkLD, # nolint: object_usage_linter.
    func_index = funcIndex, # nolint: object_usage_linter.
    sub_tuning = sub_tuning,
    lambda_vec_limit_len = lambda_vec_limit_len,
    debug_output = debug_output
  )

  return(output)
}


#' @export
gsfPEN_R <- function(
    summary_z,
    n_vec,
    plinkLD,
    func_index,
    n_iter = 1000,
    upper_val = NULL,
    breaking = 1,
    z_scale = 1,
    tuning_matrix = NULL,
    p_threshold = NULL,
    p_threshold_params = c(0.5, 10^-4, 4),
    tau_factor = c(1 / 25, 1, 3),
    len_lim_lambda = 4,
    sub_tuning = 4,
    lim_lambda = c(0.5, 0.9),
    len_lambda = 4,
    lambda_vec = NULL,
    lambda_vec_limit_len = c(1.5, 3),
    df_max = NULL,
    debug_output = FALSE) {
  time <- proc.time()

  if (z_scale != 1) {
    error("Tuning values set-up for multiple traits analysis requires z_scale=1.") # nolint: object_usage_linter.
  }

  num_func <- ncol(func_index)

  P <- nrow(summary_z)
  Q <- length(n_vec)

  summary_betas <- matrix(0, nrow = P, ncol = Q)
  SD_vec <- matrix(0, nrow = P, ncol = Q)

  for (ii in 1:Q) {
    summary_betas[, ii] <- summary_z[, ii] / sqrt(n_vec[ii])
    SD_vec[, ii] <- 1 / sqrt(n_vec[ii])
  }

  rownames(summary_betas) <- rownames(summary_z)

  if (is.null(df_max)) {
    df_max <- ceiling(0.7 * P)
  }

  if (is.null(p_threshold)) {
    p_threshold <- seq(p_threshold_params[1], p_threshold_params[2], length.out = p_threshold_params[3])
  }

  if (any(c(is.null(tuning_matrix), is.null(lambda_vec)))) {
    median_val <- median(apply(abs(summary_betas), 1, sum), na.rm = TRUE)
    tau_vec <- sort(median_val * tau_factor)
    lim_lambda <- quantile(abs(summary_z[, 1]), lim_lambda)

    output <- Tuning_setup_group_func(
      lambda_vec,
      lambda_vec_limit_len,
      p_threshold,
      num_func,
      tau_vec,
      sub_tuning,
      lim_lambda,
      len_lambda,
      len_lim_lambda,
      median_val
    )

    func_lambda <- output$func_lambda
    lambda_vec <- output$lambda_vec
    tuning_matrix <- output$tuning_matrix

    rm(output)
  } else {
    func_lambda <- permutations( # nolint: object_usage_linter.
      length(lambda_vec),
      num_func,
      repeats.allowed = TRUE
    ) - 1
  }

  beta_index <- c(seq_len(nrow(summary_betas))) - 1
  SNP_names <- rownames(summary_betas)

  # Takes only the SNPs that are present in both the PLINK data set and the GWAs
  ld_J <- PlinkLD_transform(plinkLD, SNP_names)

  J_index_matrix <- matrix(nrow = nrow(ld_J), ncol = 2)

  mat1 <- match(ld_J[, 1], SNP_names)
  mat2 <- match(ld_J[, 2], SNP_names)

  J_index_matrix[, 1] <- beta_index[mat1]
  J_index_matrix[, 2] <- beta_index[mat2]

  ld_J[, 1] <- J_index_matrix[, 1]
  ld_J[, 2] <- J_index_matrix[, 2]

  od <- order(J_index_matrix[, 1], J_index_matrix[, 2], decreasing = FALSE)
  ld_J <- ld_J[od, ]


  wind <- which(!beta_index %in% ld_J[, 1])
  num_indices <- length(wind)

  index_J <- -1
  if (num_indices > 0) {
    index_J <- beta_index[wind]
  }

  counts <- table(ld_J[, 1])
  num_SNP <- length(counts)

  index_S <- c(0, cumsum(counts)[-num_SNP])
  index_E <- cumsum(counts) - 1

  index_matrix <- matrix(nrow = num_SNP, ncol = 3)
  index_matrix[, 1] <- as.numeric(names(counts))

  index_matrix[, 2] <- index_S
  index_matrix[, 3] <- index_E

  ld_vec <- ld_J[, 3]
  ld_J <- ld_J[, 2]


  if (is.null(upper_val)) {
    upper_val <- ceiling(max(abs(summary_betas), na.rm = TRUE) * 50)
  }

  if (nrow(func_index) != nrow(summary_betas)) {
    stop("nrow of summary_betas and row of func_index do not match.")
  }


  z_matrix <- as.matrix(1 - func_index)
  # z_matrix <- matrix(1 - unlist(func_index), nrow = nrow(func_index), ncol = ncol(func_index), byrow = TRUE)
  # rownames(z_matrix) <- rownames(func_index)

  sum_func_index <- apply(z_matrix, 1, sum)
  Ifunc_SNP <- rep(0, P)
  Ifunc_SNP[which(sum_func_index != 0)] <- 1

  lambda0_vec <- abs(-qnorm(p_threshold / 2))

  nrow_index_matrix <- nrow(index_matrix)
  ncol_index_matrix <- ncol(index_matrix)

  nrow_z_matrix <- nrow(z_matrix)
  ncol_z_matrix <- ncol(z_matrix)

  nrow_func_lambda <- nrow(func_lambda)
  ncol_func_lambda <- ncol(func_lambda)

  nrow_tuning_matrix <- nrow(tuning_matrix)
  ncol_tuning_matrix <- ncol(tuning_matrix)

  nrow_all_tuning_matrix <- nrow_tuning_matrix * length(p_threshold) * nrow_func_lambda
  ncol_all_tuning_matrix <- (num_func + 1) + 2

  nrow_beta_matrix <- nrow_all_tuning_matrix
  ncol_beta_matrix <- P * Q

  dims <- c(
    num_SNP, # 1
    P, # 2
    Q, # 3
    nrow_index_matrix, # 4
    ncol_index_matrix, # 5
    nrow_z_matrix, # 6
    ncol_z_matrix, # 7
    nrow_func_lambda, # 8
    ncol_func_lambda, # 9
    nrow_tuning_matrix, # 10
    ncol_tuning_matrix, # 11
    nrow_all_tuning_matrix, # 12 == num_tuning
    ncol_all_tuning_matrix, # 13
    nrow_beta_matrix, # 14
    ncol_beta_matrix # 15
  )

  params <- c(
    upper_val, # 1
    n_iter, # 2
    breaking, # 3
    z_scale, # 4
    df_max, # 5
    length(p_threshold), # 6
    num_indices # 7
  )


  print(paste0("Number of total tuning combinations = ", nrow_all_tuning_matrix))

  Z <- gsfPEN_cpp(
    summary_betas,
    ld_J,
    index_matrix,
    index_J,
    ld_vec,
    SD_vec,
    tuning_matrix,
    lambda0_vec,
    z_matrix,
    lambda_vec,
    func_lambda,
    Ifunc_SNP,
    dims,
    params
  )

  beta_matrix <- Z$beta_matrix
  colnames(beta_matrix) <- paste0(
    rep(SNP_names, times = Q),
    ".trait",
    rep(c(1:Q), each = P)
  )

  all_tuning_matrix <- Z$all_tuning_matrix
  colnames(all_tuning_matrix) <- c(
    "lambda0",
    paste0("lambdaf", c(1:num_func)),
    "lamda2",
    "tau2"
  )

  num_iter_vec <- Z$num_iter_vec

  # Remove the tuning combinations that did not converge (correspons to -2 in num_iter_vec)
  if (!debug_output) {
    print("Removing tuning combinations that did not converge.")
    converge_index <- which(num_iter_vec > 0)
    num_iter_vec <- num_iter_vec[converge_index]
    beta_matrix <- beta_matrix[converge_index, ]
    all_tuning_matrix <- all_tuning_matrix[converge_index, ]
  }

  output <- Clean_results(
    beta_matrix = beta_matrix,
    num_iter_vec = num_iter_vec,
    all_tuning_matrix = all_tuning_matrix
  )

  time <- proc.time() - time
  print(paste0("Time elapsed: ", time[3], " seconds"))

  return(output)
}


#' @export
gsPEN_R <- function(
    summary_z,
    n_vec,
    plinkLD,
    func_index,
    n_iter = 1000,
    upper_val = NULL,
    breaking = 1,
    z_scale = 1,
    tuning_matrix = NULL,
    p_threshold = NULL,
    p_threshold_params = c(0.5, 10^-4, 4),
    tau_factor = c(1 / 25, 1, 3),
    len_lim_lambda = 4,
    sub_tuning = 4,
    lim_lambda = c(0.5, 0.9),
    len_lambda = 4,
    lambda_vec = NULL,
    lambda_vec_limit_len = c(1.5, 3),
    df_max = NULL,
    debug_output = FALSE) {
  time <- proc.time()

  if (z_scale != 1) {
    error("Tuning values set-up for multiple traits analysis requires z_scale=1.") # nolint: object_usage_linter.
  }

  num_func <- ncol(func_index)

  P <- nrow(summary_z)
  Q <- length(n_vec)

  summary_betas <- matrix(0, nrow = P, ncol = Q)
  SD_vec <- matrix(0, nrow = P, ncol = Q)

  for (ii in 1:Q) {
    summary_betas[, ii] <- summary_z[, ii] / sqrt(n_vec[ii])
    SD_vec[, ii] <- 1 / sqrt(n_vec[ii])
  }

  rownames(summary_betas) <- rownames(summary_z)

  if (is.null(df_max)) {
    df_max <- ceiling(0.7 * P)
  }

  if (is.null(p_threshold)) {
    p_threshold <- seq(p_threshold_params[1], p_threshold_params[2], length.out = p_threshold_params[3])
  }

  if (any(c(is.null(tuning_matrix), is.null(lambda_vec)))) {
    median_val <- median(apply(abs(summary_betas), 1, sum), na.rm = TRUE)
    tau_vec <- sort(median_val * tau_factor)
    lim_lambda <- quantile(abs(summary_z[, 1]), lim_lambda)

    output <- Tuning_setup_group_func(
      lambda_vec,
      lambda_vec_limit_len,
      p_threshold,
      num_func,
      tau_vec,
      sub_tuning,
      lim_lambda,
      len_lambda,
      len_lim_lambda,
      median_val
    )

    func_lambda <- output$func_lambda
    lambda_vec <- output$lambda_vec
    tuning_matrix <- output$tuning_matrix

    rm(output)
  } else {
    func_lambda <- permutations( # nolint: object_usage_linter.
      length(lambda_vec),
      num_func,
      repeats.allowed = TRUE
    ) - 1
  }

  beta_index <- c(seq_len(nrow(summary_betas))) - 1
  SNP_names <- rownames(summary_betas)

  # Takes only the SNPs that are present in both the PLINK data set and the GWAs
  ld_J <- PlinkLD_transform(plinkLD, SNP_names)

  J_index_matrix <- matrix(nrow = nrow(ld_J), ncol = 2)

  mat1 <- match(ld_J[, 1], SNP_names)
  mat2 <- match(ld_J[, 2], SNP_names)

  J_index_matrix[, 1] <- beta_index[mat1]
  J_index_matrix[, 2] <- beta_index[mat2]

  ld_J[, 1] <- J_index_matrix[, 1]
  ld_J[, 2] <- J_index_matrix[, 2]

  od <- order(J_index_matrix[, 1], J_index_matrix[, 2], decreasing = FALSE)
  ld_J <- ld_J[od, ]


  wind <- which(!beta_index %in% ld_J[, 1])
  num_indices <- length(wind)

  index_J <- -1
  if (num_indices > 0) {
    index_J <- beta_index[wind]
  }

  counts <- table(ld_J[, 1])
  num_SNP <- length(counts)

  index_S <- c(0, cumsum(counts)[-num_SNP])
  index_E <- cumsum(counts) - 1

  index_matrix <- matrix(nrow = num_SNP, ncol = 3)
  index_matrix[, 1] <- as.numeric(names(counts))

  index_matrix[, 2] <- index_S
  index_matrix[, 3] <- index_E

  ld_vec <- ld_J[, 3]
  ld_J <- ld_J[, 2]


  if (is.null(upper_val)) {
    upper_val <- ceiling(max(abs(summary_betas), na.rm = TRUE) * 50)
  }

  if (nrow(func_index) != nrow(summary_betas)) {
    stop("nrow of summary_betas and row of func_index do not match.")
  }


  z_matrix <- as.matrix(1 - func_index)
  # z_matrix <- matrix(1 - unlist(func_index), nrow = nrow(func_index), ncol = ncol(func_index), byrow = TRUE)
  # rownames(z_matrix) <- rownames(func_index)

  sum_func_index <- apply(z_matrix, 1, sum)
  Ifunc_SNP <- rep(0, P)
  Ifunc_SNP[which(sum_func_index != 0)] <- 1

  lambda0_vec <- abs(-qnorm(p_threshold / 2))

  nrow_index_matrix <- nrow(index_matrix)
  ncol_index_matrix <- ncol(index_matrix)

  nrow_z_matrix <- nrow(z_matrix)
  ncol_z_matrix <- ncol(z_matrix)

  nrow_func_lambda <- nrow(func_lambda)
  ncol_func_lambda <- ncol(func_lambda)

  nrow_tuning_matrix <- nrow(tuning_matrix)
  ncol_tuning_matrix <- ncol(tuning_matrix)

  nrow_all_tuning_matrix <- nrow_tuning_matrix * length(p_threshold) * nrow_func_lambda
  ncol_all_tuning_matrix <- (num_func + 1) + 2

  nrow_beta_matrix <- nrow_all_tuning_matrix
  ncol_beta_matrix <- P * Q

  dims <- c(
    num_SNP, # 1
    P, # 2
    Q, # 3
    nrow_index_matrix, # 4
    ncol_index_matrix, # 5
    nrow_z_matrix, # 6
    ncol_z_matrix, # 7
    nrow_func_lambda, # 8
    ncol_func_lambda, # 9
    nrow_tuning_matrix, # 10
    ncol_tuning_matrix, # 11
    nrow_all_tuning_matrix, # 12 == num_tuning
    ncol_all_tuning_matrix, # 13
    nrow_beta_matrix, # 14
    ncol_beta_matrix # 15
  )

  params <- c(
    upper_val, # 1
    n_iter, # 2
    breaking, # 3
    z_scale, # 4
    df_max, # 5
    length(p_threshold), # 6
    num_indices # 7
  )


  print(paste0("Number of total tuning combinations = ", nrow_all_tuning_matrix))

  Z <- gsfPEN_cpp(
    summary_betas,
    ld_J,
    index_matrix,
    index_J,
    ld_vec,
    SD_vec,
    tuning_matrix,
    lambda0_vec,
    z_matrix,
    lambda_vec,
    func_lambda,
    Ifunc_SNP,
    dims,
    params
  )

  beta_matrix <- Z$beta_matrix
  colnames(beta_matrix) <- paste0(
    rep(SNP_names, times = Q),
    ".trait",
    rep(c(1:Q), each = P)
  )

  all_tuning_matrix <- Z$all_tuning_matrix
  colnames(all_tuning_matrix) <- c(
    "lambda0",
    paste0("lambdaf", c(1:num_func)),
    "lamda2",
    "tau2"
  )

  num_iter_vec <- Z$num_iter_vec

  # Remove the tuning combinations that did not converge (correspons to -2 in num_iter_vec)
  if (!debug_output) {
    print("Removing tuning combinations that did not converge.")
    converge_index <- which(num_iter_vec > 0)
    num_iter_vec <- num_iter_vec[converge_index]
    beta_matrix <- beta_matrix[converge_index, ]
    all_tuning_matrix <- all_tuning_matrix[converge_index, ]
  }

  output <- Clean_results(
    beta_matrix = beta_matrix,
    num_iter_vec = num_iter_vec,
    all_tuning_matrix = all_tuning_matrix
  )

  time <- proc.time() - time
  print(paste0("Time elapsed: ", time[3], " seconds"))

  return(output)
}
